objects/src/notes/note_tag.rs

```diff
@@ -19,72 +24,17 @@ impl NoteTag {
-    /// - For network execution, the two most significant bits are set to 0b10 and the remaining
-    ///   bits are set to the 30 most significant bits of the account ID.
-    pub fn from_account_id(
-        account_id: AccountId,
-        execution: NoteExecutionMode,
-    ) -> Result<Self, NoteError> {
-        match execution {
-            NoteExecutionMode::Local => {
-                let id: u64 = account_id.into();
-                // select the 16 high bits of the account id
-                let high_bits = id & 0xffff000000000000;
-                // set bits (30,14] with the account id data
-                // set bits (32,30] as `0b00` identifying the note as intended for local execution
-                Ok(Self((high_bits >> 34) as u32))
-            },
-            NoteExecutionMode::Network => {
-                if !account_id.is_on_chain() {
-                    Err(NoteError::NetworkExecutionRequiresOnChainAccount)
-                } else {
-                    let id: u64 = account_id.into();
-                    // select the 30 high bits of the account id
-                    let high_bits = id & 0xfffffffc00000000;
-                    // set bits (30,0] with the account id data
-                    let tag = (high_bits >> 34) as u32;
-                    // set bits (32,30] as `0b10` identifying the note as intended for network
-                    // execution
-                    Ok(Self(tag | 0x80000000))
-                }
-            },
-        }
+    pub fn from_account_id(account_id: AccountId, note_type: NoteType) -> Result<Self, NoteError> {
+        let note_type_bits = (note_type as u32) << 30;
+        let account_id_highbits = (u64::from(account_id) & 0xffff000000000000) >> 34;
+        let tag = note_type_bits | (account_id_highbits as u32);
+
+        Ok(NoteTag(tag))
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    /// Returns true if the note is intended for execution by a specific account.
-    ///
-    /// A note is intended for execution by a single account if either the first two bits are zeros
-    /// or the first 3 bits are 0b100.
-    pub fn is_single_target(&self) -> bool {
-        let first_2_bit = self.0 >> 30;
-        let first_3_bits = self.0 >> 29;
-        first_2_bit == 0b00 || first_3_bits == 0b100
-    }
-
-    /// Returns note execution mode defined by this tag.
-    ///
-    /// If the most significant bit of the tag is 0 or the 3 most significant bits are equal to
-    /// 0b101, the note is intended for local execution; otherwise, the note is intended for
-    /// network execution.
-    pub fn execution_mode(&self) -> NoteExecutionMode {
-        let first_bit = self.0 >> 31;
-        let first_3_bits = self.0 >> 29;
-
-        if first_bit == 0 || first_3_bits == 0b101 {
-            NoteExecutionMode::Local
-        } else {
-            NoteExecutionMode::Network
-        }
-    }
```

Comment:
What is the reason for removing these methods? I think they provide useful info about the tag.

Reply:
How is this correct? We are only setting the highbits to the note type

Reply:
Did you mean *note tag*? The first 3 bits of the note tag define these properties. But maybe I didn't understand the question?

Reply:
That is to say, here are have tests against the 3highbits, and the kernel only checks the 2highbits.

I think my confusion is coming from all these magic values, to me they are opaque and have not meaning. It doesn't make sense to me how the network execution can be asserted in the Rust code while the Masm code has no mention to it.

Reply:
> Did you mean note tag?

Yes, sorry. I meant note tag.

> The first 3 bits of the note tag define these properties. But maybe I didn't understand the question?

I think I'm misunderstanding the objectives here. [This tables](https://github.com/0xPolygonMiden/miden-base/issues/402#issuecomment-2008740524) has lots of magic values, for example `101: local execution for a use case, public note.`, the third bit in there, is that bit the first bit of the [OFF_CHAIN storage type](https://github.com/0xPolygonMiden/miden-base/blob/dd9667d79cf3d4cc5d6a7ccc8509bdf77fdc0289/objects/src/accounts/account_id.rs#L68)? And that is intentional, not an accident?

So the validation we have here is not to guarantee that we have a valid tag, but to forbid a completely invalid one? For example, we will [`For encrypted notes: first two tag bits must be 00.`](https://github.com/0xPolygonMiden/miden-base/issues/402#issuecomment-2008740524), and at the same time we will [allow an `OffChain`](https://github.com/0xPolygonMiden/miden-base/blob/18dfadb9e5af18b932293ac085c91cdd36b390ef/miden-tx/src/tests.rs#L206) to also start with `00`?

Reply:
The way I think about this is that the first 3 bits define the properties of a tag. At the protocol level (i.e., in the kernel) we try to catch inconsistencies between tag and note type values and for this, 2 bits are enough. But we cannot guarantee that the user constructs a tag correctly.

So, for example, if the tag starts with `100`, we say that this tag is for a note intended for network execution. The only thing we can enforce here is that the note type must be `public` (this is because if the note is not public, the network cannot execute it). The way we check for public notes (according to the rules I listed above) is that the second bit must be `0` (which it is in this case). Seemingly, because we are not checking the 3rd bit, there could be an inconsistency - but this is not the case because `101` is also a tag for a public note (but intended for local execution).

Reply:
Actually, looking through the rules more closely, I do think there is an error. Specifically, here are the note types which valid for every tag prefix:

To spell the rules out a bit differently for various tag prefixes:

- tag prefix `00`: any note type works.
- tag prefix `01`: must be a private (off-chain) note.
- tag prefix `101`: must be a public note.
- tag prefix `100`: must be a public note.
- tag prefix `11`: must be a public note.

The last prefix is where the error is as the second bit is not `0`.

Reply:
So, I guess the rules should be:

- For private notes, the first bit must be `0`.
- For public notes, the first bit must be `1`.
- For encrypted notes, the first two bits must be `00`.

Reply:
> But we cannot guarantee that the user constructs a tag correctly.

Question: Why don't you want to guarantee the tag construction?

Reply:
> To spell the rules out a bit differently for various tag prefixes:

Where did these values come from? For example, given the description above:

```
For private notes, the first bit must be 0.
For encrypted notes, the first two bits must be 00.
```

It seems odd to me that `00` would pass both the validation of encrypted and private notes. And from the discussion so far it seems this is just an accident?

Reply:
> It seems odd to me that `00` would pass both the validation of encrypted and private notes. And from the discussion so far it seems this is just an accident?

No, this part is intended. Basically, `00` means that the note is intended to be executed locally by a specific recipient. This covers all note types (i.e., public, private, and encrypted notes can be executed locally).